### PR TITLE
Code LLM fine-tuning notebook

### PR DESCRIPTION
**Pull request by MKhalusova**
This notebook provides a "recipe" for fine-tuning a Code LLM on Custom Code on a single GPU. It is meant to complement the ["Personal Copilot: Train Your Own Coding Assistant" blog post](https://huggingface.co/blog/personal-copilot), which talks about data collection workflow, comparing full fine-tuning with optimized fine-tuning, and using the fine-tuned model in VS Code, and other aspects.
This notebook illustrates only the fine-tuning part and can be a valuable addition to the original article.

### CODE REVIEW

File: notebooks/Fine_tuning_Code_LLM_on_single_GPU.ipynb [hunk 1]
```diff
@@ -0,0 +1,5363 @@
+{
```
Comment (sayakpaul) [inline notebooks/Fine_tuning_Code_LLM_on_single_GPU.ipynb:1]: <p>Do we need to provide links to the models mentioned?</p><p> </p><p>Is it also worth mentioning that a meta goal of this notebook is to be able to run the entire process in a single notebook and single GPU (IIUC)? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/8/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (sayakpaul) [inline notebooks/Fine_tuning_Code_LLM_on_single_GPU.ipynb:1]: <p>Why is <code>flash-attn</code> needed here?  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/8/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (sayakpaul) [inline notebooks/Fine_tuning_Code_LLM_on_single_GPU.ipynb:1]: <p>Very nice!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/8/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (sayakpaul) [inline notebooks/Fine_tuning_Code_LLM_on_single_GPU.ipynb:1]: <blockquote>The character to token ratio of the dataset is: 2.43 </blockquote><p>Do we want to add a statement here? Like commenting the ratio if it's good or bad for the context? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/8/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (sayakpaul) [inline notebooks/Fine_tuning_Code_LLM_on_single_GPU.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    tokenizer = AutoTokenizer.from_pretrained(MODEL, use_auth_token=True, trust_remote_code=True)
</pre><p>Let's maybe not use <code>use_auth_token</code> since it's deprecated?  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/8/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (sayakpaul) [inline notebooks/Fine_tuning_Code_LLM_on_single_GPU.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #3.    def chars_token_ratio(dataset, tokenizer, data_column, nb_examples=400):
</pre><p>What does <code>nb_examples</code> stand for? Why is it set to 400? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/8/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (sayakpaul) [inline notebooks/Fine_tuning_Code_LLM_on_single_GPU.ipynb:1]: <p>A brief rundown of what we're doing in plain English would be nice. Or even adding comments in the methods would do. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/8/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (sayakpaul) [inline notebooks/Fine_tuning_Code_LLM_on_single_GPU.ipynb:1]: <p>Let's also add a short snippet for performing inference too. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/8/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='30'/>

Comment (MKhalusova) [reply notebooks/Fine_tuning_Code_LLM_on_single_GPU.ipynb:1]: <p>These are links, they just don't render weirdly in the ReviewNB app, but if you hover, you can see that they are there :) But I'll add a note on the meta goal of the notebook. </p>

Comment (MKhalusova) [reply notebooks/Fine_tuning_Code_LLM_on_single_GPU.ipynb:1]: <p>When we load the model with <code>from_pretrained</code> use it for faster training. I mention it briefly in the paragraph above. Do you think it's worth explaining more? </p><p>Now that you mention it, I notice that I should use <code>attn_implementation="flash_attention_2"</code>   instead of <code>use_flash_attention_2=True</code> which is deprecated.  </p>

Comment (MKhalusova) [reply notebooks/Fine_tuning_Code_LLM_on_single_GPU.ipynb:1]: Added a statement.

Comment (MKhalusova) [reply notebooks/Fine_tuning_Code_LLM_on_single_GPU.ipynb:1]: Added a note.

Comment (MKhalusova) [reply notebooks/Fine_tuning_Code_LLM_on_single_GPU.ipynb:1]: Added comments in the code.

Comment (MKhalusova) [reply notebooks/Fine_tuning_Code_LLM_on_single_GPU.ipynb:1]: Done!

File: notebooks/Fine_tuning_Code_LLM_on_single_GPU.ipynb [hunk 2]
```diff
@@ -0,0 +1,2338 @@
...
```
Comment (sayakpaul) [inline notebooks/Fine_tuning_Code_LLM_on_single_GPU.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #14.    base_model.cuda()
</pre><p> </p><p>not: We can avoid this added LoC if we do <code>.cuda()</code> in the above line:</p><p> </p><pre class="ql-syntax" spellcheck="false">base_model = AutoModelForCausalLM.from_pretrained(
&nbsp; &nbsp; MODEL,
&nbsp; &nbsp; quantization_config=None,
&nbsp; &nbsp; device_map=None,
&nbsp; &nbsp; trust_remote_code=True,
&nbsp; &nbsp; torch_dtype=torch.bfloat16,
).cuda()
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/cookbook/pull/8/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='33'/>

**Review by sayakpaul — APPROVED** [review]
Looking so good ❤️
